### PR TITLE
Dialog Fragment with details of an especific character added

### PR DESCRIPTION
**Pull request by julian-mainardis-globant**
## Description

Added a Dialog Fragment showing the details of the character selected

## Story Type

- New feature :white_check_mark:
- Change request
- Bug fix
- Tech

## How Can This Be Tested?

Playing the emulator and pressing the button in the main activity, then you have to scroll in the middle of the screen to see all the characters, after that, you have to click in one of them and a fragment will appear.

## Evidences

| First Header  | Second Header |
| ------------- | ------------- |
| pixel3 | nexusS |

## Coverage
image
image
image

## Checklist

- All the unit test passed :white_check_mark:

## Tested On
- Pixel 3
- Nexus S

## Package Structure
image image

### CODE REVIEW

File: app/src/main/java/com/example/marvelapp/api/service/CharacterDetailService.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.example.marvelapp.api.service
+
+import com.example.marvelapp.api.util.Result
+import com.example.marvelapp.entity.Character
+
+interface CharacterDetailService {
+    fun getCharacterById(characterId: String): Result<Character>
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/api/service/CharacterDetailService.kt]: this should be in the same MarvelService that you had created

File: app/src/main/java/com/example/marvelapp/api/service/CharacterDetailServiceImpl.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package com.example.marvelapp.api.service
+
+import com.example.marvelapp.api.MarvelApi
+import com.example.marvelapp.api.util.Result
+import com.example.marvelapp.api.util.transformToLocalCharacter
+import com.example.marvelapp.entity.Character
+
+class CharacterDetailServiceImpl(private val api: ServiceGenerator) : CharacterDetailService {
+
+    override fun getCharacterById(characterId: String): Result<Character> {
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/api/service/CharacterDetailServiceImpl.kt]: same here

File: app/src/main/java/com/example/marvelapp/activity/CharacterListActivity.kt [hunk 3]
```diff
@@ -54,7 +56,18 @@ class CharacterListActivity : AppCompatActivity(), KoinComponent {
...
+        characterDetailFragment.show(fragmentManager, TAG)
+    }
+
+    override fun setOnClickListener(characterId: String) {
+        viewModel.onCharacterPressed(characterId)
+    }
+
     companion object {
         fun getInstance(context: Context) = Intent(context, CharacterListActivity::class.java)
+        const val TAG = "CHARACTER_DETAIL_FRAGMENT"
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/activity/CharacterListActivity.kt]: this TAG should be in the fragmentDialog, not here

File: app/src/main/java/com/example/marvelapp/fragment/CharacterDetailFragment.kt [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import org.koin.android.ext.android.inject
+import org.koin.core.component.KoinComponent
+
+class CharacterDetailFragment : DialogFragment(), KoinComponent {
+
+    private lateinit var binding: FragmentCharacterDetailBinding
+    private val viewModel: CharacterDetailFragmentViewModel by inject()
+    private lateinit var characterId: String
+
+    override fun onCreate(savedInstanceState: Bundle?) {
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/fragment/CharacterDetailFragment.kt:22]: please order correctly the lines in the onCreate

File: app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharacterListViewModel.kt [hunk 5]
```diff
@@ -37,19 +37,26 @@ class CharacterListViewModel(private val model: CharacterListModel) : ViewModel(
...
         RESPONSE_SUCCESS,
         RESPONSE_ERROR,
-        RESPONSE_ERROR_NOT_FOUND
+        RESPONSE_ERROR_NOT_FOUND,
+        CHARACTER_PRESSED
     }
 
     companion object {
         private const val NOT_FOUND = "NOT_FOUND"
+        private const val EMPTY_STRING = ""
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharacterListViewModel.kt]: you have these constants in several places. Isn't better to create a ConstantUtils object and have all the constants there?

File: app/src/test/java/com/example/marvelapp/viewmodel/CharacterDetailFragmentViewModelTest.kt [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+import com.example.marvelapp.mvvm.viewmodel.CharacterListViewModel
+import com.example.marvelapp.testObserver
+import com.example.marvelapp.usecase.GetCharacterByIdUseCase
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.setMain
+import org.junit.*
```
Comment (SartiMau) [inline app/src/test/java/com/example/marvelapp/viewmodel/CharacterDetailFragmentViewModelTest.kt]: do not use wildcard imports please

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/test/java/com/example/marvelapp/viewmodel/CharacterDetailFragmentViewModelTest.kt (hunk 6 above)
Comment (julian-mainardis-globant) [reply app/src/test/java/com/example/marvelapp/viewmodel/CharacterDetailFragmentViewModelTest.kt]: done

File: app/src/main/java/com/example/marvelapp/activity/CharacterListActivity.kt (hunk 3 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/marvelapp/activity/CharacterListActivity.kt]: done

File: app/src/main/java/com/example/marvelapp/api/service/CharacterDetailService.kt (hunk 1 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/marvelapp/api/service/CharacterDetailService.kt]: done

File: app/src/main/java/com/example/marvelapp/api/service/CharacterDetailServiceImpl.kt (hunk 2 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/marvelapp/api/service/CharacterDetailServiceImpl.kt]: done

File: app/src/main/java/com/example/marvelapp/fragment/CharacterDetailFragment.kt (hunk 4 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/marvelapp/fragment/CharacterDetailFragment.kt:22]: done

File: app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharacterListViewModel.kt (hunk 5 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharacterListViewModel.kt]: done

File: app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharacterDetailFragmentViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        }
+    }
+
+    fun onButtonPressed() {
+        _characterDetailState.value = CharacterDetailData(CharacterDetailState.EXIT_BUTTON_PRESSED)
+    }
+
+    data class CharacterDetailData(
+        val state: CharacterDetailState,
+        val data: Character = Character(Constants.EMPTY_STRING, Constants.EMPTY_STRING, Constants.EMPTY_STRING, Constants.EMPTY_STRING)
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharacterDetailFragmentViewModel.kt]: in this case, character should be nulleable

File: app/src/main/java/com/example/marvelapp/util/Constants.kt [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package com.example.marvelapp.util
+
+object Constants {
+    const val NOT_FOUND = "NOT_FOUND"
+    const val EMPTY_STRING = ""
+    const val CHARACTER_ID = "CHARACTER_ID"
+    const val TAG = "CHARACTER_DETAIL_FRAGMENT"
+    const val TEST_ID = "1011334"
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/util/Constants.kt]: each test should be atomic, so each test has to have its own constants. Remove these TESTS_ contestants from here please

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/marvelapp/util/Constants.kt (hunk 8 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/marvelapp/util/Constants.kt]: done

File: app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharacterDetailFragmentViewModel.kt (hunk 7 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharacterDetailFragmentViewModel.kt]: done

File: app/src/main/java/com/example/marvelapp/api/util/Extensions.kt [hunk 9]
```diff
@@ -20,6 +21,15 @@ fun DataResponse.transformToLocalCharacterList(): List<Character> {
     return characterList
 }
 
+fun CharacterResponse.transformToLocalCharacter(): Character {
```
Comment (DiegoDiazGlobant) [inline app/src/main/java/com/example/marvelapp/api/util/Extensions.kt]: You could write this function as:

```
fun CharacterResponse.transformToLocalCharacter() =
    Character(
        this.id,
        this.name,
        this.description,
        this.image.getPathAndExtension()
    )
```

Comment (DiegoDiazGlobant) [reply app/src/main/java/com/example/marvelapp/api/util/Extensions.kt]: Same with other functions

File: app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharacterDetailFragmentViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+                        _characterDetailState.postValue(CharacterDetailData(state = CharacterDetailState.RESPONSE_ERROR_NOT_FOUND))
+                    } else {
+                        _characterDetailState.postValue(CharacterDetailData(state = CharacterDetailState.RESPONSE_ERROR))
+                    }
+                }
+            }
+        }
+    }
+
+    fun onButtonPressed() {
```
Comment (DiegoDiazGlobant) [inline app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharacterDetailFragmentViewModel.kt]: Add a more decriptive name to this function, what button?

Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharacterDetailFragmentViewModel.kt]: done

File: app/src/main/java/com/example/marvelapp/api/util/Extensions.kt (hunk 9 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/marvelapp/api/util/Extensions.kt]: done

File: app/src/main/java/com/example/marvelapp/api/util/Extensions.kt [hunk 11]
```diff
@@ -21,36 +21,31 @@ fun DataResponse.transformToLocalCharacterList(): List<Character> {
...
         this.name,
         this.description,
         this.imageURL
     )
-}
 
-fun ImageResponse.getPathAndExtension(): String {
-    return "${this.path}.${this.extension}"
-}
+fun ImageResponse.getPathAndExtension() = "${this.path}.${this.extension}"
```
Comment (DiegoDiazGlobant) [inline app/src/main/java/com/example/marvelapp/api/util/Extensions.kt]: Don't hardcode, use a DOT constant

Comment (SartiMau) [reply app/src/main/java/com/example/marvelapp/api/util/Extensions.kt]: +1

File: app/src/main/java/com/example/marvelapp/fragment/CharacterDetailFragment.kt [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+            binding.characterDetailLoader.visibility = View.GONE
+            binding.textViewCharacterDetail.visibility = View.GONE
+            binding.apply {
+                this.textViewNameCharacterDetail.text = it?.name
+                if (it?.description.isNullOrEmpty()) {
+                    this.textViewDescriptionCharacterDetail.text = resources.getString(R.string.empty_description)
+                } else {
+                    this.textViewDescriptionCharacterDetail.text = it?.description
+                }
+                Glide.with(requireContext())
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/fragment/CharacterDetailFragment.kt]: do not use `requireContext()` use context instead

File: app/src/main/java/com/example/marvelapp/api/util/Extensions.kt (hunk 11 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/marvelapp/api/util/Extensions.kt]: done

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/marvelapp/fragment/CharacterDetailFragment.kt (hunk 12 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/marvelapp/fragment/CharacterDetailFragment.kt]: done

**Review by DiegoDiazGlobant — APPROVED** [review]

**Review by kearneyeduardo — APPROVED** [review]

**Review by kearneyeduardo — APPROVED** [review]

**Review by SartiMau — APPROVED** [review]

**Pull request merged**